feat(tianmu): tianmu exporter support NULL type. (#1169)

lujiashun opened the pull request:
<!--

Thank you for contributing to StoneDB!

PR Title Format: <type>(<scope>): description to this pr (#issue_id)
e.g.
fix(util): fix sth..... (#1)

-->

## Summary about this PR
<!--

There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".
e.g.:
Issue: close #1

-->

Issue Number: close #1169

1 fix sql load when varchar column's char length(should be byte length) exceeds input-text byte length; 
2 support user varialbe "TIANMU_LOAD_NULL" to customize the export file content; 
3 fix the NULL value of exporter when enclosed char is not empty;

## Tests Check List
<!-- At least one of next options must be included. -->

- [ ] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog
<!-- At least one of next options must be included. -->

- [x] New Feature
- [ ] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation
<!-- At least one of next options must be included. -->

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

RingsC commented on storage/tianmu/loader/parsing_strategy.cpp:
```diff
@@ -606,9 +606,9 @@ void ParsingStrategy::GetValue(const char *value_ptr, size_t value_size, ushort
   } else if (core::ATI::IsTxtType(ati.Type())) {
     // process escape characters
-    if (ati.CharLen() < (uint)value_size) {
+    if (ati.Precision() < (uint)value_size) {
```
1: using c++ style type cast. 
2: how about the boundary checking, such as ati.Precision() == value_size.

lujiashun replied:
1. Ok, I will fix it.
2.  (ati.Precision() < (uint)value_size) means the input text exceeds the max length,  (ati.Precision() == value_size) is available

hustjieke commented on storage/tianmu/exporter/data_exporter_txt.cpp:
```diff
@@ -92,14 +92,14 @@ void DEforTxt::PutRowEnd() {
 }
 
-size_t DEforTxt::WriteString(const types::BString &str, int len, bool text_or_bin) {
+size_t DEforTxt::WriteString(const types::BString &str, int len, bool text_or_bin, bool isNull) {
```
Change `isNull` to `is_null`.
```
The names of variables (including function parameters) and data members are all lowercase, with underscores between words. See:
```
https://google.github.io/styleguide/cppguide.html#Variable_Names

lujiashun replied:
OK, I will fix it.

RingsC commented on storage/tianmu/core/engine.h:
```diff
@@ -313,6 +313,7 @@ class ResultExportSender final : public ResultSender {
 };
 
 enum class tianmu_var_name {
+  TIANMU_TIMEOUT,
```
The first enum item should be set a default value.

lujiashun replied:
OK, I will fix it;

hustjieke commented on storage/tianmu/exporter/data_exporter_txt.cpp:
```diff
@@ -92,14 +92,14 @@ void DEforTxt::PutRowEnd() {
   // If you specify OPTIONALLY, the ENCLOSED BY character is used only to enclose values from columns that
   // have a string data type (such as CHAR, BINARY, TEXT, or ENUM):
-  if (str_qualifier_ && (text_or_bin || !opt_enclosed_)) {
+  if (str_qualifier_ && !isNull && (text_or_bin || !opt_enclosed_)) {
```
The same problem as above.

lujiashun replied:
OK, I will change it to follow Google code style;

Commit:
feat(tianmu): tianmu exporter support NULL type. (#1169)

[summary]
1 fix sql load when varchar column's char length(should be byte length) exceeds input-text byte length;
2 support user varialbe "TIANMU_LOAD_NULL" to customize the export file content;
3 fix the NULL value of exporter when enclosed char is not empty;

RingsC commented on storage/tianmu/exporter/data_exporter_txt.h:
```diff
@@ -46,9 +46,9 @@ class DEforTxt : public DataExporter {
   }
   void WriteDelimiter() { WriteChars(delimiter_); }
-  void WriteNull() { WriteString(nulls_str_.c_str(), (int)nulls_str_.length()); }
+  void WriteNull() { WriteString(nulls_str_.c_str(), (int)nulls_str_.length(), false, true); }
```
same as: c++ styple type cast

lujiashun replied:
OK, I will fix it;

RingsC reviewed (approved):
nice, LGTM

duanfuxiang0 reviewed (approved):
LGTM

mergify[bot] committed:
Merge branch 'stonedb-5.7-dev' into feat-1169-stonedb5.7
